### ISSUE TITLE
User remains authenticated after logout in other tabs until refresh

### ISSUE THREAD

**Issue by ajahangard** [oqtane/oqtane.framework#3108]
Hi
I think there is a small security problem in implementation of Authentication.
If I open a browser and login with admin account. then I open another tab in browser I am logged in as administrator too.
If I logout in one of the tabs, I will have administrator access in other tab until refresh or reload.
I think that's because the current circuit is not notified that the user has already logged out.
Maybe using RevalidatingServerAuthenticationStateProvider to check AuthenticationState periodically solves this problem.

**Comment by meisam-dehghan** [comment]
Nice point,Here's what Steve Anderson commented on a similar issue.Hope that would help:

> UserManager.UpdateSecurityStampAsync(userId) is the way to cause a user's existing logins to be invalidated. You can change how often your application reacts to this by editing the RevalidatingIdentityAuthenticationStateProvider.cs file in your project. The default RevalidationInterval value has it check for expiry every 30 minutes.

> If you're looking for some kind of instantly synchronised sign-out, that's not a feature we have. You could build it yourself by having a singleton DI service that raises events whenever anyone signs out, and have some code running inside each circuit listen to that event and sign the user out if you see the same user signed out in a different circuit. But it gets more complicated if you have multiple servers, since they would have independent state. That's the whole reason why, in the end, we default to using RevalidatingIdentityAuthenticationStateProvider with some RevalidationInterval to trigger polling the underlying identity database for security stamp changes.


https://github.com/dotnet/aspnetcore/issues/18901#issuecomment-584255080

**Comment by sbwalker** [comment]
@ajahangard you are correct that each browser tab has its own dedicated circuit when running on Blazor Server. There are a few options I will explore. Oqtane has a synchronization event service which may handle this scenario. Or it could be handled in the IdentityAuthenticationService

**Comment by meisam-dehghan** [comment]
@sbwalker  Well,I have already written a custom authentication provider inheriting from AuthenticationStateProvider,but as far as I'm concerned,the problem is that once a user logs out from a tab,the components running within only that circut will be notified of the change in the authentication state.The user in other tabs will remain authenticated as long as he/she doesn't somehow refresh the page by either manually hitting the refresh button in the browser or calling the NavigateTo method with the reload parameter set to true.

**Comment by ajahangard** [comment]
@sbwalker I think the synchronization event is a good start point to explore too.

**Comment by sbwalker** [comment]
@ajahangard PR #3114 handles the concern you raised. It uses the eventing system to notify the router that the user has logged out - which then triggers a page reload to refresh the circuit. As part of investigating this scenario I also tested some other user auth scenarios which should trigger a reload - user password change, change in role assignments for user, etc... Note that these changes are recognized immediately and are handled on the next UI navigation in the client application - they do not rely on a timer (which is what the RevalidatingServerAuthenticationStateProvider uses).
